### ISSUE TITLE
[Doc-Clarity] torch.scatter with duplicated indices

### ISSUE THREAD

**Issue by Enigmatisms** [pytorch/pytorch#162242]
### 📚 The doc issue

In the document, it is stated that:

> When indices are not unique, the behavior is non-deterministic (one of the values from src will be picked arbitrarily) and the gradient will be incorrect (it will be propagated to all locations in the source that correspond to the same index)!

However, I could not reproduce non-deterministic behavior, the results look rather deterministic. Check the following minimal reproducing example:

```python
import numpy as np
import torch


def test_non_deterministic():
    gt_res = np.array(
        [
            [
                [-99.0, 1.0, -97.0, 3.0],
                [-96.0, -98.0, 6.0, 7.0],
                [8.0, 9.0, 10.0, 11.0],
            ],
            [
                [-90.0, -92.0, 14.0, 15.0],
                [-93.0, 17.0, -91.0, 19.0],
                [20.0, 21.0, 22.0, 23.0],
            ],
        ],
        dtype=np.float32,
    )
    gt_input_grad = np.array(
        [
            [[0.0, 1.0, 0.0, 1.0], [0.0, 0.0, 1.0, 1.0], [1.0, 1.0, 1.0, 1.0]],
            [[0.0, 0.0, 1.0, 1.0], [0.0, 1.0, 0.0, 1.0], [1.0, 1.0, 1.0, 1.0]],
        ],
        dtype=np.float32,
    )
    gt_src_grad = np.array(
        [
            [[1.0, 1.0, 1.0], [1.0, 1.0, 1.0]],
            [[1.0, 1.0, 1.0], [1.0, 1.0, 1.0]],
        ],
        dtype=np.float32,
    )
    for i in range(1000):
        torch.cuda.empty_cache()
        inputs = torch.arange(24, dtype=torch.float32).reshape([2, 3, 4]).cuda()
        src = torch.arange(-99, -99 + (2 * 2 * 3), dtype=torch.float32).reshape(
            [2, 2, 3]
        ).cuda()
        index = torch.tensor(
            [
                [
                    [0, 1, 0],
                    [1, 1, 0],
                ],
                [
                    [1, 0, 1],
                    [0, 0, 1],
                ],
            ],
            dtype=torch.int64,
        ).cuda()

        inputs.requires_grad = True
        src.requires_grad = True

        res = torch.scatter(inputs, 1, index, src)
        res.backward(torch.ones_like(res))

        print(f"Test {i + 1}/{1000}")
        np.testing.assert_allclose(res.cpu().detach().numpy(), gt_res)
        np.testing.assert_allclose(inputs.grad.cpu().numpy(), gt_input_grad)
        np.testing.assert_allclose(src.grad.cpu().numpy(), gt_src_grad)


if __name__ == "__main__":
    test_non_deterministic()
``` 
The test is run on both A100 and A800 with torch `2.8.0+cu128`, and I can't get an AssertionError as expected.

There are four main questions:

- Question 1: If the results are truly non-deterministic, why would I be able to use one single set of GTs for testing? I tried to avoid caching by emptying the cache and redefine the variables (1000 times), but it does not seem to work.
- Question 2: The pattern for `torch.scatter` seems pretty obvious: if there are duplicated indices, we **use the element with the smallest indices on the scatter axis**. Am I correct about this? At least the input grad and result suggest so. Should it be clarified in the doc?
- Question 3: What's wrong with src grad? If the above pattern is correct (taking the element with smallest on-dim indices), the correct src grad should actually be:

```python
[
  [
    [1, 1, 1],
    [1, 0, 0],
  ],
  [
    [1, 1, 1],
    [1, 0, 0],
  ],
]
```

instead of a `ones`. I, somewhat, can comprehend whats going on under the hood: all the elements in src are traversed (accessed once) but some of them are overwritten so they should nevertheless get their grads updated? But this seems so unreasonable. I wonder if this is just a intended feature but not recommended for use, or a bug. If it is a intended feature, then `torch.scatter` should be deterministic, but it isn't so I guess it is a bug.

- Question 4: CPU behavior is just the opposite to that of GPU's. It selects the element with the largest indices. The src grad is, expectedly incorrect (all 1), and it is also not deterministic. You can try that by deleting the `.cuda()` in the code above. I would doubt that CPU is implemented via atomic ops, so it won't be non-deterministic inherently, therefore src grad should be easy to calculate correctly. 

### Suggest a potential alternative/fix

For doc: it would be better to clarify the behavior (whether it is truly deterministic or not; why result/input grad are correct while src grad isn't; CPU behavior clarification). Maybe you can lead me (us) to the source code of how PyTorch implements the forward and backward pass of `torch.scatter`.

**Comment by ngimel** [comment]
If you tested on something larger than a toy example, you'd see that results are actually non-deterministic. 
```
import torch

a = torch.randn(8192, device="cuda")
src = torch.arange(4096, device="cuda", dtype=torch.float)
ind = torch.randint(128, (4096,), device="cuda")

res = torch.scatter(a, 0, ind, src)

for i in range(100):
    tmp = torch.scatter(a, 0, ind, src)
    if (tmp-res).abs().max() != 0:
        print(i, "results differ!")

```
We have no ability to determine which of the src values was actually picked, thus all the src values corresponding to the same index receive a grad, that's what's meant by "gradients will be incorrect".

**Comment by Enigmatisms** [comment]
Thanks for elaborating.
